File: 827. Making A Large Island/main.go
package main

func largestIsland(grid [][]int) int {
	n := len(grid)

	outOfBounds := func(r, c int) bool {
		return r < 0 || c < 0 || r == n || c == n
	}

	var dfs func(r, c, label int) int
	dfs = func(r, c, label int) int {
		if outOfBounds(r, c) || grid[r][c] != 1 {
			return 0
		}
		grid[r][c] = label
		size := 1
		nei := [][2]int{{r - 1, c}, {r + 1, c}, {r, c - 1}, {r, c + 1}}
		for _, item := range nei {
			nr, nc := item[0], item[1]
			size += dfs(nr, nc, label)
		}
		return size
	}

	size := make(map[int]int)
	label := 2
	for r := range n {
		for c := range n {
			if grid[r][c] == 1 {
				size[label] = dfs(r, c, label)
				label++
			}
		}
	}

	connect := func(r, c int) int {
		visit := make(map[int]bool)
		nei := [][2]int{{r - 1, c}, {r + 1, c}, {r, c - 1}, {r, c + 1}}
		res := 1
		for _, item := range nei {
			nr, nc := item[0], item[1]
			if outOfBounds(nr, nc) || grid[nr][nc] == 0 || visit[grid[nr][nc]] {
				continue
			}
			res += size[grid[nr][nc]]
			visit[grid[nr][nc]] = true
		}
		return res
	}

	res := 1
	for r := range n {
		for c := range n {
			if grid[r][c] == 0 {
				res = max(res, connect(r, c))
			} else {
				res = max(res, size[grid[r][c]])
			}
		}
	}

	return res
}

File: 827. Making A Large Island/main.py
# total sam

from typing import List


class Solution:
    def largestIsland(self, grid: List[List[int]]) -> int:
        # find all islands
        # for each island cell replace 1 with island size
        # for each 0 cell look 4 directions, sum up numbers
        # create map coord -> island_id to avoid counting same island multiple times

        n = len(grid)
        directions = [(-1, 0), (0, -1), (1, 0), (0, 1)]
        coord_to_island = {}
        island_id = 0

        # bfs find island, replace all one's with island's size
        def mark_island(r, c):
            nonlocal island_id
            visit = set([(r, c)])
            q = [(r, c)]
            island_size = 0
            while q:
                r, c = q.pop()
                island_size += 1
                for dr, dc in directions:
                    row, col = r + dr, c + dc
                    if row not in range(n) or col not in range(n) or (row, col) in visit or grid[row][col] == 0:
                        continue
                    q.append((row, col))
                    visit.add((row, col))
            for r, c in visit:
                grid[r][c] = island_size
                coord_to_island[(r, c)] = island_id
            island_id += 1
        
        for r in range(n):
            for c in range(n):
                if grid[r][c] == 1:
                    mark_island(r, c)
        
        res = 1

        # get potential island size for 0 cell 
        def get_potential_island_size(r, c):
            res = 1
            included_islands = set()
            for dr, dc in directions:
                row, col = r + dr, c + dc
                if row not in range(n) or col not in range(n) or grid[row][col] == 0: continue
                cur_island = coord_to_island[(row, col)]
                if cur_island in included_islands: continue
                res += grid[row][col]
                included_islands.add(cur_island)    
            return res
        
        for r in range(n):
            for c in range(n):
                if grid[r][c] == 0:
                    res = max(res, get_potential_island_size(r, c))
                else:
                    res = max(res, grid[r][c])    

        return res
